asserts/pool.go

```diff
@@ -521,11 +525,29 @@ func (p *Pool) Add(a Assertion, grouping Grouping) (ok bool, err error) {
 // Errors related to the assertions are associated with the relevant groups
 // and can be retrieved with Err, in which case ok set to false.
 func (p *Pool) AddBatch(b *Batch, grouping Grouping) (ok bool, err error) {
-	// XXX parse the grouping only once
+	if err := p.phase(poolPhaseAdd); err != nil {
+		return false, err
+	}
+
+	// b dealt with unsupported formats already
+
+	// deserialize grouping if needed only once
+	var cachedGrouping *internal.Grouping
+	deser := func(_ string) (*internal.Grouping, error) {
+		if cachedGrouping != nil {
+			// do a copy as addToGrouping and resolveWith
+			// might add to their input
+			g := cachedGrouping.Copy()
+			return &g, nil
+		}
+		var err error
+		cachedGrouping, err = p.groupings.Deserialize(string(grouping))
+		return cachedGrouping, err
```

Comment:
Nitpick, these 3 lines could be reduced to `return p.groupings.Deserialize(string(grouping))`?

Reply:
no, is setting cachedGrouping from line 535, otherwise it would stay nil and there would be no caching

Reply:
Ah indeed, I missed that, thanks.